paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/postgres/PostgresRecordParser.java

```diff
@@ -225,24 +225,15 @@ private DataType extractFieldType(
             case "boolean":
                 return DataTypes.BOOLEAN();
             case "string":
+                int newLength = afterData.get(field.field()).asText().length();
                 if (paimonField == null) {
-                    return DataTypes.VARCHAR(afterData.get(field.field()).asText().length());
+                    return DataTypes.VARCHAR(newLength);
                 } else if (paimonField.type() instanceof VarCharType) {
                     int oldLength = ((VarCharType) paimonField.type()).getLength();
-                    int newLength = afterData.get(field.field()).asText().length();
-                    if (oldLength < newLength) {
-                        return DataTypes.VARCHAR(newLength);
-                    } else {
-                        return DataTypes.VARCHAR(oldLength);
-                    }
+                    return DataTypes.VARCHAR(Math.max(oldLength, newLength));
                 } else if (paimonField.type() instanceof CharType) {
                     int oldLength = ((CharType) paimonField.type()).getLength();
-                    int newLength = afterData.get(field.field()).asText().length();
-                    if (oldLength < newLength) {
-                        return DataTypes.CHAR(newLength);
-                    } else {
-                        return DataTypes.CHAR(oldLength);
-                    }
+                    return DataTypes.CHAR(Math.max(oldLength, newLength));
```

Comment:
in case "string", We cannot convert the varchar(n)/char(n) type into the string type, which will cause schemaCompatible to fail if the field changes when restarting the paimon-cdc job, because the string length of paimon is integer.max is always greater than n.
At the same time, when pg table performs column type alter(example: n changes from 10 to 20), because pg-debezium does not have ddl event, we can only get the length value n from the record